File: music_extractaudioclips.hs
import Text.Read
import Data.List.Split --splitOn
import System.Process
import Control.Monad
--import Data.Char --isDigit
--import System.Directory --doesFileExist
import System.IO --openFile for reading ANSI unicode characters
import Data.String.Utils --replace
--import Text.RegexPR
--import Data.List --intercalate
--import GHC.Stack
--import Debug.Trace
import GHC.IO.Encoding (setLocaleEncoding) --setLocaleEncoding, so I can write a freaking EM dash — to a file
unicodeReadFile :: String -> IO String
unicodeReadFile myfile = do
  handle <- openFile myfile ReadMode
  hSetEncoding handle utf8
  contents <- hGetContents handle
  return contents

main = do
  setLocaleEncoding utf8
  clipFilesFromSpreadSheet "in" "out" "audiolist.txt"
  return()

--Extract Audio Clips in batch, using a spreadsheet pasted into a text file
----------------
--HOW IT WORKS--
----------------
--Calls a batch of sox commands to extract clips from audio file outputs.
--the text file input is a tab-delemited list with format (including header row)
  --input file  start time  end time  output clip name (use.ogg)
  --input.mp3  1:30:01.333 1:32:03 clipname.ogg
--the trim command called is like
  --sox "myinputfile.mp3" "myoutputclip.ogg" trim 12.5 =25.333
--UPDATE: just use ffmpeg. This has the added advantage of it's possible to add a switch to clip video, too.

clipFilesFromSpreadSheet :: String -> String -> String -> IO()
clipFilesFromSpreadSheet infoldername outfoldername myclipsfilename = do
  mytrackcliplist <- readTrackSheet_toSongDataList myclipsfilename
  --putStrLn $ show $ mytrackcliplist!!0
  --mapM_ sox_clip_track mytrackcliplist
  mapM_ (\x -> ffmpeg_clip_track infoldername outfoldername x) mytrackcliplist
  return()
sox_clip_track :: SongClipData -> IO()
sox_clip_track mysongdata = do
  mycommand <- return $ "sox \""++(filename mysongdata)++"\" \""++(outputtrackname mysongdata)++"\" trim "++(starttime mysongdata)++" ="++(endtime mysongdata)
  callCommand mycommand
  return()
--2017-10-27-1925-07 Because I want to extract a song from a 4-hour video and didn't want to convert it to mp3 first; that takes a while!
ffmpeg_clip_track :: String -> String -> SongClipData -> IO()
ffmpeg_clip_track infoldername outfoldername mysongdata = do
  --mycommand <- return $ "ffmpeg -ss " ++ starttime mysongdata ++ " -i \""++(filename mysongdata)++"\" -c copy -t \""++(outputtrackname mysongdata)++"\" trim "++(starttime mysongdata)++" ="++(endtime mysongdata)
    --https://superuser.com/questions/138331/using-ffmpeg-to-cut-up-video
    --can implement hybrid seeking. if seconds is more than 60, then
  mycommand <- return $ setcodeccommand infoldername outfoldername mysongdata
  --mycommand <- return $ "ffmpeg -ss " ++ hybridseektime ++ " -i \""++(filename mysongdata)++"\" -ss " ++ accurateseekstarttime ++ " -t " ++ songduration ++ " -c:a libvorbis \""++(outputtrackname mysongdata)++"\""
  --mycommand <- return $ "ffmpeg -ss " ++ hybridseektime ++ " -i \""++(filename mysongdata)++"\" -ss " ++ accurateseekstarttime ++ " -t " ++ songduration ++ " "++mycodecstring++" \""++(outputtrackname mysongdata)++"\""
  --putStrLn mycommand
  --getLine
  callCommand mycommand
  return()
setcodeccommand :: String -> String -> SongClipData -> String
setcodeccommand infoldername outfoldername mysongdata
  -- | outextension == "ogg" = "ffmpeg -ss " ++ hybridseektime ++ " -i \""++(filename mysongdata)++"\" -ss " ++ accurateseekstarttime ++ " -t " ++ songduration ++ " "++"c:a libvorbis"++" \""++(outputtrackname mysongdata)++"\""
  -- | outextension == "mp4" = "-c:v libx264 -c:a copy"
  -- | otherwise = "-c:a libvorbis"
  | outextension == "ogg" = "ffmpeg -i \""++(infoldername++"\\"++filename mysongdata)++"\" -ss " ++ (hybridseektime mysongdata) ++ " -i \""++(infoldername++"\\"++filename mysongdata)++"\" -ss " ++ (accurateseekstarttime mysongdata) ++ " -t " ++ (songduration mysongdata) ++ " "++"-map 0:0 -map 1:1? -c:a libvorbis -qscale:v 7"++" \""++(outfoldername++"\\"++outputtrackname mysongdata)++"\""
    --what was a command that copies the cover image with seeking?
    --ffmpeg -i "a.mp3" -ss 0 -i "a.mp3" -ss 2 -t 2 -c:a libvorbis -c:v libtheora -qscale:v 7 "z.ogg"
      --breaks
    --ffmpeg -i "a.mp3" -ss 0 -i "a.mp3" -t 2 -c:a libvorbis -c:v libtheora -qscale:v 7 "z.ogg"
      --works. the hybrid seek time seems to break oggs with embedded images. Whatever.
  | outextension == "mp4" = "ffmpeg -ss " ++ (hybridseektime mysongdata) ++ " -i \""++(infoldername++"\\"++filename mysongdata)++"\" -ss " ++ (accurateseekstarttime mysongdata) ++ " -t " ++ (songduration mysongdata) ++ " "++"-c:v libx264 -c:a copy"++" \""++(outfoldername++"\\"++outputtrackname mysongdata)++"\""
  | otherwise = error $ "use mp4 or ogg extension for output"
  where
  outextension = head $ reverse $ splitOn "." $ outputtrackname mysongdata
starttime_asnum :: SongClipData -> Float
starttime_asnum mysongdata = (read $ starttime mysongdata)::Float
hybridseektime :: SongClipData -> String
hybridseektime mysongdata
  | starttime < 60 = "0"
  | otherwise = show $ starttime - 60
  where
  starttime = starttime_asnum mysongdata
accurateseekstarttime :: SongClipData -> String
accurateseekstarttime mysongdata
  | (starttime_asnum mysongdata) < 60 = show $ starttime_asnum mysongdata
  | otherwise = "60"
songduration :: SongClipData -> String
songduration mysongdata = show $ ((read $ endtime mysongdata)::Float) - ((read $ starttime mysongdata)::Float)

data SongClipData = SongClipData
  {filename::String
  ,starttime::String
  ,endtime::String
  ,outputtrackname::String
  } deriving Show
--READ DATA FROM SPREADSHEET
readTrackSheet_toSongDataList :: String -> IO([SongClipData])
readTrackSheet_toSongDataList myfilename = do
  mytrackfile <- readFile myfilename
  mytrackfile <- return $ drop 1 (splitOn "\n" mytrackfile)
  --putStr $ show mytrackfile
  mytrackfile <- return $ map (\x -> splitOn "\t" x) mytrackfile
  --putStr $ show mytrackfile
  mytrackfile <- return $ map (\x -> splitSheet_toSongClipData x) mytrackfile
  --putStr $ show mytrackfile
  return mytrackfile
splitSheet_toSongClipData :: [String] -> SongClipData
splitSheet_toSongClipData mylist
  | length mylist == 4 = (makesongclipdata mylist)
  | otherwise = makesongclipdata ["","","",""]
  where
  makesongclipdata mylist = SongClipData
    {filename=(mylist!!0)
    ,starttime=hhmmss_toSeconds $ (mylist!!1)
    ,endtime=hhmmss_toSeconds $ (mylist!!2)
    ,outputtrackname=nameoutputfile mylist
    }
  nameoutputfile mylist = (mylist!!0)++" "++(replace ":" "m" (mylist!!1))++"s "++(mylist!!3)
--[][][]PROCESS START/END TIME DATA WHEN COLUMNS ARE BLANK
--CONVERT mm:ss to just seconds
hhmmss_toSeconds :: String -> String
hhmmss_toSeconds mytimestamp
  | length stamparray == 3 = (show $ (readint_pos stamparray 0)*3600 + (readint_pos stamparray 1)*60 + (readint_pos stamparray 2)) ++ (maybemilliseconds mytimestamp 2)
  | length stamparray == 2 = (show $ (readint_pos stamparray 0)*60 + (readint_pos stamparray 1)) ++ (maybemilliseconds mytimestamp 1)
  | length stamparray == 1 = (show $ (readint_pos stamparray 0)) ++ (maybemilliseconds mytimestamp 0)
  | otherwise = "0"
  where
  maybemilliseconds :: String -> Int -> String
  maybemilliseconds mytimestamp pos
    | (length $ splitOn "." $ (splitOn ":" (mytimestamp))!!pos) == 2 = "." ++ (splitOn "." $ splitOn ":" (mytimestamp)!!pos)!!1
    | otherwise = ""
  readint_pos mynumarray pos = read (head $ splitOn "." (mynumarray!!pos))::Int
  stamparray = splitOn ":" mytimestamp

File: music_extractaudioclips.py
'''
Usage: clipFilesFromSpreadSheet(filename)

where the file is a text file with header:
file_name	start_time	end_time	output_suffix
filename.mp4    hh:mm:ss.zzz    ss.zzz  cool clip.mp4

(hh:mm:ss.zzz, mm:ss.zzz, and ss.zzz are all valid formats. zzz means milliseconds.)

Put full video or audio files in folder "in", run, and clips will be put in folder "out"; these paths can be edited. Calls FFMPEG to do the clipping.
'''

from subprocess import call

def clipFilesFromSpreadSheet(filename = "audiolist.txt"):
    #filename = "audiolist.txt"
    myfile = readFile(filename)
    myfile = myfile.split("\n")
    myfile = [x.split("\t") for x in myfile]
    myfile = myfile[1:] #exclude header row
    clipdatalines = [readClipData_fromLine(x) for x in myfile]
    for i in range(len(clipdatalines)):
        executeVideoAudioClip(clipdatalines[i])
    return
#
def readFile(myfile):
    try:
        f = open(myfile, mode="r", encoding="utf-8")
    except:
        f = open(myfile, mode="r")
    content = f.read()
    f.close()
    return content
def readClipData_fromLine(myline):
    input_video_title = myline[0]
    clip_starttime_asstring = hhmmss_toSeconds(myline[1])
    clip_endtime_asstring = hhmmss_toSeconds(myline[2])
    output_name_suffix = myline[3]
    return [input_video_title, clip_starttime_asstring, clip_endtime_asstring, output_name_suffix]
def hhmmss_toSeconds(mytimestamp):
    #valid input formats: like hh:mm:ss.zzz, mm:ss.zzz, ss, ss.zzz
    #get milliseconds
    mytimestamp_seconds = 0.
    if len(mytimestamp.split(".")) > 2:
        print("more than two periods in timestamp. that's a weird error case.")
        return None
    elif len(mytimestamp.split(".")) > 1:
        mytimestamp_seconds += float('0.'+mytimestamp.split(".")[1])
        mytimestamp = mytimestamp.split(".")[0]
    #get hours, minutes, seconds
    splitstamp = mytimestamp.split(":")
    if len(splitstamp) > 3:
        print("more than two colons in timestamp. that's a weird error case.")
        return None
    elif len(splitstamp) == 3:
        mytimestamp_seconds += 3600*float(splitstamp[0]) + 60*float(splitstamp[1]) + float(splitstamp[2])
    elif len(splitstamp) == 2:
        mytimestamp_seconds += 60*float(splitstamp[0]) + float(splitstamp[1])
    elif len(splitstamp) == 1:
        mytimestamp_seconds += float(splitstamp[0])
    return str(mytimestamp_seconds)
#
def executeVideoAudioClip(clipdata):
    input_video_title, clip_starttime_asstring, clip_endtime_asstring, output_name_suffix = clipdata
    output_extension = output_name_suffix.split(".")[-1]
    input_video_path = "in" + "/" + input_video_title
    output_video_path = "out" + "/" + makeOutputClipName(input_video_title, output_name_suffix, clip_starttime_asstring)
    #audioexts = ["ogg", "mp3", "wav"] #maybe handle these separately.
    audioexts = ["ogg"]
    videoexts = ["mp4"]
    if output_extension in audioexts:
        mycommand = "ffmpeg -i \""+(input_video_path)+"\" -ss " + (getHybridSeekTime(clip_starttime_asstring)) + " -i \""+(input_video_path)+"\" -ss " + (getAccurateStartSeekTime(clip_starttime_asstring)) + " -t " + (getClipDuration(clip_starttime_asstring, clip_endtime_asstring)) + " "+"-map 0:0 -map 1:1? -c:a libvorbis -qscale:v 7"+" \""+(output_video_path)+"\""
            #uses two inputs to try and get a cover image for audio files?... I forgot.
    elif output_extension in videoexts:
        mycommand = "ffmpeg -ss " + (getHybridSeekTime(clip_starttime_asstring)) + " -i \""+(input_video_path)+"\" -ss " + (getAccurateStartSeekTime(clip_starttime_asstring)) + " -t " + (getClipDuration(clip_starttime_asstring, clip_endtime_asstring)) + " "+"-c:v libx264 -c:a copy"+" \""+(output_video_path)+"\""
    else:
        print("use mp4 or ogg extension for the clip name.")
        return None
    call(mycommand, shell = True)
    return
def getHybridSeekTime(starttime):
    if float(starttime) < 60:
        return '0'
    else:
        return str(float(starttime) - 60)
def getAccurateStartSeekTime(starttime):
    if float(starttime) < 60:
        return starttime
    else:
        return '60'
def getClipDuration(starttime, endtime):
    return str(round(float(endtime) - float(starttime), 3))
def makeOutputClipName(input_video_title, output_name_suffix, clip_starttime_asstring):
    splitstamp = clip_starttime_asstring.split(":")
    if len(splitstamp) > 3:
        print("more than two colons in timestamp. that's a weird error case.")
        return None
    elif len(splitstamp) == 3:
        timestamp_thing = splitstamp[0]+"h"+splitstamp[1]+"m"+splitstamp[2]+"s"
    elif len(splitstamp) == 2:
        timestamp_thing = splitstamp[0]+"m"+splitstamp[1]+"s"
    elif len(splitstamp) == 1:
        alltime = float(splitstamp[0])
        if alltime < 60:
            timestamp_thing = str(alltime)+"s"
        else: #let's format this so it's human-readable
            timestamp_thing = nicelyFormatSeconds_tohhmmss(alltime)
    return input_video_title + " " + timestamp_thing + " " + output_name_suffix
def nicelyFormatSeconds_tohhmmss(secondsonly_time_asfloat):
    hours = int(secondsonly_time_asfloat // 3600)
    minutes = int((secondsonly_time_asfloat - 3600*hours)//60)
    seconds = secondsonly_time_asfloat - 3600*hours - 60*minutes
    seconds_whole = int(seconds)
    milliseconds = round(seconds - seconds_whole, 3)
    timestamp_string = ""
    if hours > 0:
        timestamp_string += str(hours).zfill(2) + "h"
    if not(hours == 0 and minutes == 0):
        timestamp_string += str(minutes).zfill(2) + "m"
    timestamp_string += str(seconds_whole).zfill(2)
    if milliseconds != 0:
        timestamp_string += "." + str(milliseconds)[2:].ljust(3,'0')
    timestamp_string += "s"
    return timestamp_string

if __name__ == "__main__":
    clipFilesFromSpreadSheet()
